platform/linux-generic/odp_schedule_scalable.c

```diff
@@ -950,18 +950,19 @@ static int _schedule(odp_queue_t *from, odp_event_t ev[], int num_evts)
 	for (i = 0, first = NULL; i < ts->num_schedq; i++, first = NULL) {
 		sched_queue_t *schedq = ts->schedq_list[i];
 		sched_elem_t *elem;
+		time_stamp = _odp_cpu_global_time();
 restart_same:
 		elem = schedq_peek(schedq);
 		if (odp_unlikely(elem == NULL)) {
 			/* Schedq empty, look at next one. */
 			continue;
 		}
 		if (is_pktin(elem)) {
-			/* Pktio ingress queue */
-			if (first == NULL)
-				first = elem;
-			else if (elem == first) /* Wrapped around */
+			if (elem->loop_check[odp_cpu_id()] == time_stamp) { /* Wrapped around */
```

Comment:
Scheduling elements are currently mostly read-only, so adding a write operation for each element (plus `CONFIG_NUM_CPU` x 8B extra data) adds a considerable overhead. On the other hand, I'm finding it hard to figure out a solution with less overhead. Would adding an internal define for the max number of pktio queues to poll per scheduling round be enough to avoid most of the problems you have described?

Reply:
hi @matiaselo 
     because ODP don't have   DP NUMBER and DP core id design.   so use CONFIG_NUM_CPU   ,maybe waste some memory. But it is very a little,relative to pkt_priv a big memory waste.  it is not a big thing .
unsigned char ODP_ALIGNED_CACHE pkt_priv[PKTIO_PRIVATE_SIZE]; //great memory waste is here .      you can ignore schedule elem .

Reply:
maybe we can use unsigned __u16 to implement the same feature,   use a thread variable to replace __u64 type timestamp .

Reply:
> because ODP don't have DP NUMBER and DP core id design. so use CONFIG_NUM_CPU ,maybe waste some memory. But it is very a little,relative to pkt_priv a big memory waste. it is not a big thing .

The size of sched_elem_t is multiplied with queue count, but anyway that is a secondary issue. What I'm worried about is writing to each `sched_elem_t` element during the scheduling loop and causing additional cache line invalidations. Currently element data is only read if no packets are received.

Reply:
One option would be to always schedule pktin queues using the original scheduling queue and not move empty queues to a lower priority sched queue. This would improve QoS seen by the input queues but decrease max throughput if not all input queues are receiving traffic.

Reply:
for hw queues, we can not know which one have traffic with high priority, sometime no traffic , other time  many traffic, it is normal status in network . we depend on dynamical priority.     ignore my patch3 uint16 which is not perfect to avoid loop.  uint64 is still perfect .

regarding to  cacheline , we can test the real performance difference , is it a big thing , to check. in my test setup,  the performance is hightest value after using my fix .

Reply:
Isn't it so, that the dead lock actually happens when multiple pktio queue elements have already been "demoted" to the lower priority idle scheduling queue? When looping through this idle queue and some other thread moves the 'first' element we end up in a dead lock. However, in case of a normal priority scheduling queue the idle pktio queue elements will always be demoted to the lower priority queue and the loop will eventually break since there are no more queue elements left in the current scheduling queue.

Now, would it be acceptable to simply add a define (or dynamically calculated value) to set the maximum number of "idle" pktio queue elements to poll per scheduling loop? This would minimize the overhead seen by normal priority queues.

Reply:
Thinking more about it, adding one more if clause (`if (elem->schedq != schedq)`) removes most of the overhead for the non-idle pktin queues.

E.g.
```
if (is_pktin(elem)) {
	/* Check idle pktin queue for wraparound */
	if (elem->schedq != schedq) {
		int thread_id = odp_thread_id();

		if (elem->loop_check[thread_id] == loop_cnt)
			continue;
		elem->loop_check[thread_id] = loop_cnt;
	}
...
```